--- a/init.luau	2026-06-01 01:43:01.231746549 +0100
+++ b/init.luau	2026-06-01 01:35:56.391293960 +0100
@@ -131,33 +131,8 @@
             i = self.start
             while index < length and i < 64 do
                 code = message:byte(index+1)
-                if code < 0x80 then
-                    blocks[rshift(i, 2)] = lshift(code, shift[band(i, 3)])
-                    i+=1
-                elseif code < 0x800 then
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0xc0, rshift(code, 6)), shift[band(i, 3)]))
-                    i+=1
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(code, 0x3f)), shift[band(i, 3)]))
-                    i+=1
-                elseif code < 0xd800 or code >= 0xe000 then
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0xe0, rshift(code, 12)), shift[band(i, 3)]))
-                    i+=1
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(rshift(code, 6), 0x3f)), shift[band(i, 3)]))
-                    i+=1
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(code, 0x3f)), shift[band(i, 3)]))
-                    i+=1
-                else
-                    code = 0x10000 + bor(lshift(band(code, 0x3ff), 10), band( message:byte(index+1), 0x3ff))
-                    i+=1
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0xf0, rshift(code, 18)), shift[band(i, 3)]))
-                    i+=1
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(rshift(code, 12), 0x3f)), shift[band(i, 3)]))
-                    i+=1
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(rshift(code, 6), 0x3f)), shift[band(i, 3)]))
-                    i+=1
-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(code, 0x3f)), shift[band(i, 3)]))
-                    i+=1
-                end
+                blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(code, shift[band(i, 3)]))
+                i+=1
                 index+=1
             end
         end
@@ -214,11 +189,7 @@

 function sha256.hash(self)
     local function to_signed32(n)
-        if n >= 0 then
-            return n
-        else
-            return n + 4294967296
-        end
+        return bit32.band(n, 0xFFFFFFFF)
     end

     local a = self.h0
