export type CacheType = "Application"
	| "DiscordEmoji"
	| "DiscordGuilded"
	| "DiscordGuildRole"
	| "DiscordUser"
	| "DiscordChannel"
	| "DiscordTextChannel"
	| "DiscordVoiceChannel"
	| "DiscordMessage"
	| "DiscordMember"
	| "DiscordInvite"
	| "AutoModeration"

return {
	Application = "Application",
	DiscordEmoji = "DiscordEmoji",
	DiscordGuild = "DiscordGuild",
	DiscordGuildRole = "DiscordGuildRole",
	DiscordUser = "DiscordUser",
	DiscordChannel = "DiscordChannel",
	DiscordTextChannel = "DiscordTextChannel",
	DiscordVoiceChannel = "DiscordVoiceChannel",
	DiscordMessage = "DiscordMessage",
	DiscordMember = "DiscordMember",
	DiscordInvite = "DiscordInvite",
	AutoModeration = "AutoModeration",
}