return function()
	local lines = {
		"diff --git a/init.luau b/init.luau",
		"index 0ab7093..30328fa 100644",
		"--- a/init.luau",
		"+++ b/init.luau",
		"@@ -131,33 +131,8 @@ function sha256.update(self, message)",
		"             i = self.start\r",
		"             while index < length and i < 64 do\r",
		"                 code = message:byte(index+1)\r",
		"-                if code < 0x80 then\r",
		"-                    blocks[rshift(i, 2)] = lshift(code, shift[band(i, 3)])\r",
		"-                    i+=1\r",
		"-                elseif code < 0x800 then\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0xc0, rshift(code, 6)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(code, 0x3f)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                elseif code < 0xd800 or code >= 0xe000 then\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0xe0, rshift(code, 12)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(rshift(code, 6), 0x3f)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(code, 0x3f)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                else\r",
		"-                    code = 0x10000 + bor(lshift(band(code, 0x3ff), 10), band( message:byte(index+1), 0x3ff))\r",
		"-                    i+=1\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0xf0, rshift(code, 18)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(rshift(code, 12), 0x3f)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(rshift(code, 6), 0x3f)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                    blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(bor(0x80, band(code, 0x3f)), shift[band(i, 3)]))\r",
		"-                    i+=1\r",
		"-                end\r",
		"+                blocks[rshift(i, 2)] = bor(blocks[rshift(i, 2)], lshift(code, shift[band(i, 3)]))\r",
		"+                i+=1\r",
		"                 index+=1\r",
		"             end\r",
		"         end\r",
		"@@ -214,11 +189,7 @@ end",
		" \r",
		" function sha256.hash(self)\r",
		"     local function to_signed32(n)\r",
		"-        if n >= 0 then\r",
		"-            return n\r",
		"-        else\r",
		"-            return n + 4294967296\r",
		"-        end\r",
		"+        return bit32.band(n, 0xFFFFFFFF)\r",
		"     end\r",
		" \r",
		"     local a = self.h0\r",
	}

	return table.concat(lines, "\n") .. "\n"
end
