export type MessageType =
	"Default"
	| "RecipientAdd"
	| "RecipientRemove"
	| "Call"
	| "ChannelNameChange"
	| "ChannelIconChange"
	| "ChannelPinnedMessage"
	| "UserJoin"
	| "GuildBoost"
	| "GuildBoostTier1"
	| "GuildBoostTier2"
	| "GuildBoostTier3"
	| "ChannelFollowAdd"
	| "GuildDiscoveryDisqualified"
	| "GuildDiscoveryRequalified"
	| "GuildDiscoveryGracePeriodInitialWarning"
	| "GuildDiscoveryGracePeriodFinalWarning"
	| "ThreadCreated"
	| "Reply"
	| "ChatInputCommand"
	| "ThreadStarterMessage"
	| "GuildInviteReminder"
	| "ContextMenuCommand"
	| "AutoModerationAction"
	| "RoleSubscriptionPurchase"
	| "InteractionPremiumUpsell"
	| "StageStart"
	| "StageEnd"
	| "StageSpeaker"
	| "StageTopic"
	| "GuildApplicationPremiumSubscription"
	| "GuildIncidentAlertModeEnabled"
	| "GuildIncidentAlertModeDisabled"
	| "GuildIncidentReportRaid"
	| "GuildIncidentReportFalseAlarm"
	| "PurchaseNotification"
	| "PollResult"

export type MessageAcitvityType = "Join" | "Spectate" | "Listen" | "JoinRequest"
export type MessageReferenceType = "Default" | "Forward"

export type InteractionType =
	"Ping"
	| "ApplicationCommand"
	| "MessageComponent"
	| "ApplicationCommandAutocomplete"
	| "ModalSubmit"

export type ButtonStyle = "Blurple" | "Grey" | "Green" | "Red" | "Grey"

export type SelectDefaultValueType = "User" | "Role" | "Channel"

export type TextInputStyle = "Short" | "Paragraph"

return "DISCORD_V10_MESSAGE_TYPES"
